File: src/main/java/com/artillexstudios/axminions/command/AxMinionsCommand.java
package com.artillexstudios.axminions.command;

import com.artillexstudios.axapi.utils.ContainerUtils;
import com.artillexstudios.axapi.utils.MessageUtils;
import com.artillexstudios.axapi.utils.logging.LogUtils;
import com.artillexstudios.axminions.AxMinionsPlugin;
import com.artillexstudios.axminions.config.Config;
import com.artillexstudios.axminions.config.Language;
import com.artillexstudios.axminions.config.Minions;
import com.artillexstudios.axminions.config.Skins;
import com.artillexstudios.axminions.minions.Level;
import com.artillexstudios.axminions.minions.Minion;
import com.artillexstudios.axminions.minions.MinionArea;
import com.artillexstudios.axminions.minions.MinionData;
import com.artillexstudios.axminions.minions.MinionType;
import com.artillexstudios.axminions.minions.MinionWorldCache;
import com.artillexstudios.axminions.utils.Direction;
import com.artillexstudios.axminions.utils.FileUtils;
import com.artillexstudios.axminions.utils.LocationUtils;
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import revxrsal.commands.annotation.Command;
import revxrsal.commands.annotation.Default;
import revxrsal.commands.annotation.Optional;
import revxrsal.commands.annotation.Subcommand;
import revxrsal.commands.bukkit.annotation.CommandPermission;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicInteger;

@Command({"axminions", "minions", "minion"})
public final class AxMinionsCommand {

    @Subcommand("give")
    @CommandPermission("axminions.command.give")
    public void give(CommandSender sender, Player player, MinionType type, @Optional Level level, @Optional @Default("1") Integer amount) {
        handleGive(sender, player, type, level, amount);
    }

    @Subcommand("version")
    @CommandPermission("axminions.command.version")
    public void version(CommandSender sender) {
        MessageUtils.sendMessage(sender, Language.PREFIX, "<green>You are running <white>AxMinions</white> version <white><version></white> on <white><implementation></white> version <white><implementation-version></white> (Implementing API version <white><api-version></white>)",
                Placeholder.unparsed("version", AxMinionsPlugin.instance().getDescription().getVersion()),
                Placeholder.unparsed("implementation", Bukkit.getName()),
                Placeholder.unparsed("implementation-version", Bukkit.getVersion()),
                Placeholder.unparsed("api-version", Bukkit.getBukkitVersion())
        );
    }

    @Subcommand("reload")
    @CommandPermission("axminions.command.reload")
    public void reload(CommandSender sender) {
        long start = System.nanoTime();
        List<File> failed = new ArrayList<>();

        for (World world : Bukkit.getWorlds()) {
            MinionWorldCache.clear(world);
        }

        if (!Config.reload()) {
            failed.add(FileUtils.PLUGIN_DIRECTORY.resolve("config.yml").toFile());
        }

        if (!Language.reload()) {
            failed.add(FileUtils.PLUGIN_DIRECTORY.resolve("language").resolve(Language.lastLanguage + ".yml").toFile());
        }

        if (!Skins.reload()) {
            failed.add(FileUtils.PLUGIN_DIRECTORY.resolve("skins.yml").toFile());
        }

        Minions.reload();
        failed.addAll(Minions.failedToLoad());

        CompletableFuture<?>[] futures = Minions.loadingMinions().toArray(new CompletableFuture[0]);
        List<World> worlds = Bukkit.getWorlds();
        AtomicInteger counter = new AtomicInteger();
        CompletableFuture.allOf(futures).thenRun(() -> {
            Minions.loadingMinions().clear();

            for (World world : worlds) {
                AxMinionsPlugin.instance().handler().loadMinions(world).toCompletableFuture().thenAccept(loaded -> {
                    if (Config.debug) {
                        LogUtils.debug("Loaded {} minions in world {} in {} ms!", loaded.firstInt(), world.getName(), loaded.secondLong() / 1_000_000);
                    }
                    if (counter.incrementAndGet() != futures.length * worlds.size()) {
                        return;
                    }

                    if (failed.isEmpty()) {
                        MessageUtils.sendMessage(sender, Language.PREFIX, Language.RELOAD_SUCCESS, Placeholder.parsed("time", Long.toString((System.nanoTime() - start) / 1_000_000)));
                    } else {
                        MessageUtils.sendMessage(sender, Language.PREFIX, Language.RELOAD_FAIL, Placeholder.parsed("time", Long.toString((System.nanoTime() - start) / 1_000_000)), Placeholder.parsed("files", String.join(", ", failed.stream()
                                .map(File::getName)
                                .toList())
                        ));
                    }
                });
            }
        });
    }

    @Subcommand("debug")
    @CommandPermission("axminions.command.debug")
    public static final class Debug {

        @Subcommand("spawn")
        @CommandPermission("axminions.command.debug.spawn")
        public void spawn(Player sender, MinionType type, Level level) {
            Location location = LocationUtils.toBlockCenter(sender.getLocation());
            MinionData data = new MinionData(0, type, Direction.NORTH, null, level, 0, new ItemStack(Material.DIAMOND_PICKAXE), null, new HashMap<>());
            Minion minion = new Minion(0, location, data);
            minion.spawn();
            MinionArea area = MinionWorldCache.getArea(location.getWorld());
            MinionWorldCache.add(minion);
            area.startTicking(location.getChunk());

        }
    }

//    public static void register(JavaPlugin plugin) {

//        new CommandTree("axminions")
//                .withAliases("minion", "minions")
//                .then(new LiteralArgument("give")
//                        .withPermission("axminions.command.give")
//                        .then(new EntitySelectorArgument.OnePlayer("player")
//                                .then(MinionTypeArgument.minionType("miniontype")
//                                        .then(MinionLevelArgument.level("level")
//                                                .then(new IntegerArgument("amount")
//                                                        .executes((sender, args) -> {
//                                                            Player player = args.getByClass("player", Player.class);
//                                                            MinionType type = args.getByClass("miniontype", MinionType.class);
//                                                            Level level = args.getByClass("level", Level.class);
//                                                            Integer amount = args.getByClass("amount", Integer.class);
//                                                            handleGive(sender, player, type, level, amount);
//                                                        })
//                                                )
//                                        )
//                                )
//                        )
//                )
//                .then(new LiteralArgument("version")
//                        .withPermission("axminions.command.version")
//                        .executes((sender, args) -> {
//                            MessageUtils.sendMessage(sender, Language.PREFIX, "<green>You are running <white>AxMinions</white> version <white><version></white> on <white><implementation></white> version <white><implementation-version></white> (Implementing API version <white><api-version></white>)",
//                                    Placeholder.unparsed("version", AxMinionsPlugin.instance().getDescription().getVersion()),
//                                    Placeholder.unparsed("implementation", Bukkit.getName()),
//                                    Placeholder.unparsed("implementation-version", Bukkit.getVersion()),
//                                    Placeholder.unparsed("api-version", Bukkit.getBukkitVersion())
//                            );
//                        })
//                )
//                .then(new LiteralArgument("reload")
//                        .withPermission("axminions.command.reload")
//                        .executes((sender, args) -> {
//                            long start = System.nanoTime();
//                            List<File> failed = new ArrayList<>();
//
//                            for (World world : Bukkit.getWorlds()) {
//                                MinionWorldCache.clear(world);
//                            }
//
//                            if (!Config.reload()) {
//                                failed.add(FileUtils.PLUGIN_DIRECTORY.resolve("config.yml").toFile());
//                            }
//
//                            if (!Language.reload()) {
//                                failed.add(FileUtils.PLUGIN_DIRECTORY.resolve("language").resolve(Language.lastLanguage + ".yml").toFile());
//                            }
//
//                            if (!Skins.reload()) {
//                                failed.add(FileUtils.PLUGIN_DIRECTORY.resolve("skins.yml").toFile());
//                            }
//
//                            Minions.reload();
//                            failed.addAll(Minions.failedToLoad());
//
//                            CompletableFuture<?>[] futures = Minions.loadingMinions().toArray(new CompletableFuture[0]);
//                            List<World> worlds = Bukkit.getWorlds();
//                            AtomicInteger counter = new AtomicInteger();
//                            CompletableFuture.allOf(futures).thenRun(() -> {
//                                Minions.loadingMinions().clear();
//
//                                for (World world : worlds) {
//                                    AxMinionsPlugin.instance().handler().loadMinions(world).toCompletableFuture().thenAccept(loaded -> {
//                                        if (Config.debug) {
//                                            LogUtils.debug("Loaded {} minions in world {} in {} ms!", loaded.firstInt(), world.getName(), loaded.secondLong() / 1_000_000);
//                                        }
//                                        if (counter.incrementAndGet() != futures.length * worlds.size()) {
//                                            return;
//                                        }
//
//                                        if (failed.isEmpty()) {
//                                            MessageUtils.sendMessage(sender, Language.PREFIX, Language.RELOAD_SUCCESS, Placeholder.parsed("time", Long.toString((System.nanoTime() - start) / 1_000_000)));
//                                        } else {
//                                            MessageUtils.sendMessage(sender, Language.PREFIX, Language.RELOAD_FAIL, Placeholder.parsed("time", Long.toString((System.nanoTime() - start) / 1_000_000)), Placeholder.parsed("files", String.join(", ", failed.stream()
//                                                    .map(File::getName)
//                                                    .toList())
//                                            ));
//                                        }
//                                    });
//                                }
//                            });
//                        })
//                )
//                .then(new LiteralArgument("debug")
//                        .withPermission("axminions.command.debug")
//                        .then(new LiteralArgument("spawn")
//                                .withPermission("axminions.command.debug.spawn")
//                                .then(MinionTypeArgument.minionType("miniontype")
//                                        .then(MinionLevelArgument.level("level")
//                                                .executesPlayer((sender, args) -> {
//                                                    MinionType type = args.getByClass("miniontype", MinionType.class);
//                                                    Level level = args.getByClass("level", Level.class);
//                                                    Location location = LocationUtils.toBlockCenter(sender.getLocation());
//                                                    MinionData data = new MinionData(0, type, Direction.NORTH, null, level, 0, new ItemStack(Material.DIAMOND_PICKAXE), null, new HashMap<>());
//                                                    Minion minion = new Minion(0, location, data);
//                                                    minion.spawn();
//                                                    MinionArea area = MinionWorldCache.getArea(location.getWorld());
//                                                    MinionWorldCache.add(minion);
//                                                    area.startTicking(location.getChunk());
//                                                })
//                                        )
//                                )
//                        )
//                )
//                .register();
//    }
//
    private static void handleGive(CommandSender sender, Player player, MinionType type, Level level, Integer amount) {
        ItemStack item = type.item(new MinionData(0, type, null, null, level, 0, null, null, new HashMap<>(0)));
        item.setAmount(amount);
        ContainerUtils.INSTANCE.addOrDrop(player.getInventory(), List.of(item), player.getLocation());
    }
}

File: common/src/main/kotlin/com/artillexstudios/axminions/commands/AxMinionsCommand.kt
package com.artillexstudios.axminions.commands

import com.artillexstudios.axapi.utils.ItemBuilder
import com.artillexstudios.axapi.utils.StringUtils
import com.artillexstudios.axminions.AxMinionsPlugin
import com.artillexstudios.axminions.api.AxMinionsAPI
import com.artillexstudios.axminions.api.config.Config
import com.artillexstudios.axminions.api.config.Messages
import com.artillexstudios.axminions.api.minions.miniontype.MinionType
import com.artillexstudios.axminions.api.minions.miniontype.MinionTypes
import com.artillexstudios.axminions.api.utils.fastFor
import com.artillexstudios.axminions.converter.LitMinionsConverter
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder
import org.bukkit.OfflinePlayer
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import revxrsal.commands.annotation.*
import revxrsal.commands.bukkit.annotation.CommandPermission

@Command("axminions", "minion", "minions")
class AxMinionsCommand {

    @Subcommand("give")
    @CommandPermission("axminions.command.give")
    @Description("Give a minion to a player")
    @AutoComplete("* @minionTypes * *")
    fun give(
        sender: CommandSender,
        receiver: Player,
        minionType: MinionType,
        @Default("1") level: Int,
        @Default("1") @Range(min = 1.0, max = 64.0) amount: Int
    ) {
        val item = minionType.getItem(level)
        item.amount = amount

        receiver.inventory.addItem(item)
    }

    @Subcommand("fuel give")
    @CommandPermission("axminions.command.fuel.give")
    fun give(sender: CommandSender, id: String, receiver: Player, @Default("1") amount: Int) {
        val item = ItemBuilder(Config.CHARGE_ITEMS().getSection(id)).get()
        item.amount = amount
        receiver.inventory.addItem(item)
    }

    @Subcommand("fuel add")
    @CommandPermission("axminions.command.fuel.add")
    fun add(player: Player, id: String, charge: Int) {
        val item = ItemBuilder(player.inventory.itemInMainHand).serialize(true)
        item["charge"] = charge
        AxMinionsPlugin.config.getConfig().set("charge.items.$id", item)
        AxMinionsPlugin.config.getConfig().save()
    }

    @Subcommand("reload")
    @CommandPermission("axminions.command.reload")
    @Description("Reload the configurations of the plugin")
    fun reload(sender: CommandSender) {
        val start = System.currentTimeMillis()
        AxMinionsPlugin.config.reload()
        AxMinionsPlugin.messages.reload()

        MinionTypes.getMinionTypes().fastFor { _, v ->
            v.getConfig().reload()
        }

        AxMinionsAPI.INSTANCE.getMinions().fastFor {
            it.markDirty()
        }

        AxMinionsPlugin.integrations.reload()

        sender.sendMessage(
            StringUtils.formatToString(
                Messages.PREFIX() + Messages.RELOAD_SUCCESS(),
                Placeholder.unparsed("time", (System.currentTimeMillis() - start).toString())
            )
        )
    }

    @Subcommand("convert")
    @CommandPermission("axminions.command.convert")
    @Description("Convert from a different plugin")
    fun convert(sender: CommandSender) {
        val converter = LitMinionsConverter()
        converter.convert()
    }

    @Subcommand("stats", "statistics")
    @CommandPermission("axminions.command.statistics")
    @Description("Get statistics of plugin")
    fun stats(sender: CommandSender) {
        val minions = AxMinionsAPI.INSTANCE.getMinions()
        var loaded = 0
        val total = minions.size

        minions.fastFor {
            if (it.isTicking()) {
                loaded++
            }
        }

        sender.sendMessage(
            StringUtils.formatToString(
                Messages.STATISTICS(),
                Placeholder.unparsed("ticking", loaded.toString()),
                Placeholder.unparsed("not-ticking", (total - loaded).toString()),
                Placeholder.unparsed("total", total.toString())
            )
        )
    }

    @Subcommand("reset")
    @CommandPermission("axminions.command.reset")
    fun reset(sender: CommandSender, offlinePlayer: OfflinePlayer) {
        val minions = AxMinionsAPI.INSTANCE.getMinions()
        val ownerUUID = offlinePlayer.uniqueId
        AxMinionsPlugin.dataQueue.submit {
            minions.fastFor {
                if (it.getOwnerUUID() == ownerUUID) {
                    it.remove()
                }
            }

            sender.sendMessage(StringUtils.formatToString(Messages.PREFIX() + Messages.RESET(), Placeholder.unparsed("player", offlinePlayer.name ?: "---")))
        }
    }

    @Subcommand("extraslot")
    @CommandPermission("axminions.command.extraslot")
    fun extraSlot(commandSender: CommandSender, offlinePlayer: OfflinePlayer, amount: Int) {
        AxMinionsPlugin.dataQueue.submit {
            val original = AxMinionsPlugin.dataHandler.getExtraSlots(offlinePlayer.uniqueId)
            AxMinionsPlugin.dataHandler.addUser(offlinePlayer.uniqueId, offlinePlayer.name ?: "---")
            AxMinionsPlugin.dataHandler.addExtraSlot(offlinePlayer.uniqueId, amount)
            commandSender.sendMessage(StringUtils.formatToString(Messages.PREFIX() + Messages.SLOT_GIVE(), Placeholder.unparsed("player", offlinePlayer.name ?: "???"), Placeholder.unparsed("amount", amount.toString())))
            offlinePlayer.player?.sendMessage(StringUtils.formatToString(Messages.PREFIX() + Messages.SLOT_RECEIVE(), Placeholder.unparsed("amount", amount.toString()), Placeholder.unparsed("from", original.toString()), Placeholder.unparsed("to", (original + amount).toString())))
        }
    }
}